File: run_replays.py
from __future__ import annotations

import argparse
import json
from pathlib import Path

from replay.runner import run_replay_file
from runtime_backend import (
    backend_argument_help,
    exit_backend_missing,
    normalize_backend,
    run_rust_bin,
    rust_backend_available,
)


ROOT = Path(__file__).resolve().parent
FLOWS = ROOT / "schemas" / "examples" / "flows"


def run_python() -> int:
    summaries = []
    for flow_path in sorted(FLOWS.glob("*.json")):
        result = run_replay_file(flow_path)
        summaries.append(
            {
                "replay_id": result.fixture.replay_id,
                "events": len(result.event_log),
                "task_terminal": result.state.current_task_terminal(result.fixture.task_id),
                "verified_facts": [fact_id for fact_id, fact in sorted(result.state.facts.items()) if fact["status"] == "verified"],
                "approvals": len(result.state.approvals),
                "recoveries": len(result.state.recoveries),
            }
        )
    print(json.dumps(summaries, indent=2, ensure_ascii=False))
    return 0


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--backend", choices=("auto", "python", "rust"), default="auto", help=backend_argument_help())
    args = parser.parse_args()

    backend = normalize_backend(args.backend)
    if backend == "rust":
        if not rust_backend_available():
            return exit_backend_missing()
        return run_rust_bin("run_replays")
    return run_python()


if __name__ == "__main__":
    raise SystemExit(main())

File: rust-core/src/bin/run_replays.rs
use std::fs;
use std::path::PathBuf;

use device_agent_core::{FactStatus, ReplayRunner};
use serde_json::json;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let flows_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("schemas")
        .join("examples")
        .join("flows");

    let mut flow_paths = fs::read_dir(&flows_dir)?
        .filter_map(|entry| entry.ok().map(|item| item.path()))
        .filter(|path| path.extension().and_then(|value| value.to_str()) == Some("json"))
        .collect::<Vec<_>>();
    flow_paths.sort();

    let mut summaries = Vec::new();
    for flow_path in flow_paths {
        let mut runner = ReplayRunner::from_path(&flow_path)?;
        let result = runner.run(None)?;
        let task_id = format!("task_{}", result.fixture.replay_id);
        let task_terminal = result
            .state
            .task_states
            .get(&task_id)
            .map(|state| serde_json::to_value(state).ok())
            .flatten()
            .and_then(|value| value.as_str().map(ToOwned::to_owned))
            .unwrap_or_else(|| "created".to_owned());
        let verified_facts = result
            .state
            .facts
            .iter()
            .filter(|(_, fact)| fact.status == FactStatus::Verified)
            .map(|(fact_id, _)| fact_id.clone())
            .collect::<Vec<_>>();

        summaries.push(json!({
            "replay_id": result.fixture.replay_id,
            "events": result.event_log.len(),
            "task_terminal": task_terminal,
            "verified_facts": verified_facts,
            "approvals": result.state.approvals.len(),
            "recoveries": result.state.recoveries.len(),
        }));
    }

    println!("{}", serde_json::to_string_pretty(&summaries)?);
    Ok(())
}
